### ISSUE TITLE
How should the ingress gateway be called from javascript?

### ISSUE THREAD

**Issue by Shige99011** [hashicorp/consul-helm#950]
Hi, I'm using Ingress gateway to make it access from an external service that is simple javascript application. When an API that is in consul service mesh is called via Ingress gateway, an error like below happens (Actually I got the error as I mentioned in [#946](https://github.com/hashicorp/consul-helm/issues/946). So I temporarily disabled the CORS policy of my browser).

`hello.js:11 GET http://192.168.202.199:31000/hello-service/hello 404 (Not Found)`

The javascript code is:
```

    fetch('http://192.168.202.199:31000/hello-service/hello', {
        method: 'GET',
        headers: {
            'Host': 'api-service.ingress.consul', //This header is going to be ignored?
        }
      })
      .then(response => {
          response.text()
          .then(text => {
              alert(text)
            })
        })
        .catch(error => {
            alert(error);
        });

```
I think the request header is going to be ignored for fetch() as the header key is "Host". Is it correct? If so, how should the ingress gateway be called from javascript?

The response is returned correctly when it is called by using curl command like:
`curl -H "Host: api-service.ingress.consul" "http://192.168.202.199:31000/hello-service/hello"`

My values.yaml:
```
global:
  name: consul

server:
  replicas: 1
client:
  enabled: true
connectInject:
  enabled: true
controller:
  enabled: true
  
terminatingGateways:
  enabled: true
  defaults:
    replicas: 1
  gateways:
    - name: terminating-gateway

ingressGateways:
  enabled: true
  defaults:
    replicas: 1
  gateways:
    - name: ingress-gateway
      service:
        type: NodePort
        ports:
          - port: 7000
            nodePort: 31000
```

Thanks for your help.

**Comment by ishustava** [comment]
Hey @Shige99011 

Could you tell us how you've configured your ingress gateway and the api-service?

**Comment by Shige99011** [comment]
Thanks @ishustava for your reply.

My configuration for ingress gateway is

```
apiVersion: consul.hashicorp.com/v1alpha1
kind: IngressGateway
metadata:
  name: ingress-gateway
spec:
  listeners:
    - port: 7000
      protocol: http
      services:
        - name: api-service
```

And api-service:
```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: api-service
---
apiVersion: v1
kind: Pod
metadata:
  name: api-service
  annotations:
    'consul.hashicorp.com/connect-inject': 'true'
spec:
  containers:
    - name: api-service
      image: api-service:alpha3
      imagePullPolicy: IfNotPresent 
      ports:
        - containerPort: 10000
          name: http
      env:
        - name: 'LISTEN_PORT'
          value: '10000'
    # If ACLs are enabled, the serviceAccountName must match the Consul service name.
  serviceAccountName: api-service
```

Also, routing configuration  has been applied as below:
```
apiVersion: consul.hashicorp.com/v1alpha1
kind: ServiceRouter
metadata:
  name: api-service
spec:
  routes:
    - match:
        http:
          pathPrefix: "/hello-service"
      destination:
        service: "hello-service"
    - match:
        http:
          pathPrefix: "/konnichiwa-service"
      destination:
        service: "konnichiwa-service"
    - match:
        http:
          pathPrefix: "/external-service"
      destination:
        service: "external-service"
```

 And service default is here:

```
apiVersion: consul.hashicorp.com/v1alpha1
kind: ServiceDefaults
metadata:
  name: api-service
spec:
  protocol: http
---

apiVersion: consul.hashicorp.com/v1alpha1
kind: ServiceDefaults
metadata:
  name: hello-service
spec:
  protocol: http
---
apiVersion: consul.hashicorp.com/v1alpha1
kind: ServiceDefaults
metadata:
  name: konnichiwa-service
spec:
  protocol: http
---

apiVersion: consul.hashicorp.com/v1alpha1
kind: ServiceDefaults
metadata:
  name: external-service
spec:
  protocol: http
```

btw, terminating gateway is also used:

```
apiVersion: consul.hashicorp.com/v1alpha1
kind: TerminatingGateway
metadata:
  name: terminating-gateway
spec:
  services:
    - name: external-service
```

The register information for terminating gateway is:

```
{
  "Node": "External Service",
  "Address": "192.168.49.23",
  "NodeMeta": {
    "external-node": "true",
    "external-probe": "true"
  },
  "Service": {
    "ID": "external-service",
    "Service": "external-service",
    "Port": 55002
  }
}
```


My environment is below:
OS: ubuntu 18.04 on hyper-v
microk8s: v1.20.5
kubernetes: v1.20.5
consul: v1.9.4
Consul helm chart: v0.31.1
(Consul helm chart was installed by command "microk8s helm3 install -f config.yaml consul hashicorp/consul --version "0.31.1"")
There is only one kubernetes node.


Thanks for your help.

**Comment by ishustava** [comment]
Thanks so much for this info! 

I'm sorry, I have not noticed in your original message that when you curl directly, it works. I'm not an expert on javascript to be honest, and don't know if I can help beyond what's required on the consul side. There's nothing specific to how ingress gateway works with javascript or plain curl, and so I think it's just the matter of figuring out the right code.

Looking at the documentation for the [`fetch()`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#headers) function, it looks like some headers, including the `Host` header, are [forbidden](https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name). Maybe that is your issue? However, I cannot provide you with the javascript code that you need as this is a little outside of my area of expertise.

**Comment by Shige99011** [comment]
Thanks for the reply.
yes, my concern is the 'Host' header that is in forbidden list. But 'Host' header is required to call the API via ingress gateway, right?

Anyway, I will have a look more from javascript perspective.

Thank you.

**Comment by ishustava** [comment]
> yes, my concern is the 'Host' header that is in forbidden list. But 'Host' header is required to call the API via ingress gateway, right?

Correct, the `Host` header is required. That is how ingress gateway know which service to route to.

**Comment by Shige99011** [comment]
This issue has been solved in #954 .
